### DISCUSSION TITLE
能否支持音乐内嵌封面或歌词

### DISCUSSION THREAD

**Discussion by ChirmyRam** [AlistGo/alist#230] [Ideas]
1. 我注意到在文件列表中可以显示音乐内嵌封面，但播放器只能用后台设定的封面，我的音乐基本都有内嵌封面，能否在播放时显示每首歌自己的内嵌封面？
2. 更进一步，我的音乐也基本都内嵌歌词，能否支持显示内嵌歌词？
![image](https://user-images.githubusercontent.com/34760505/146641552-7439ba11-152f-423a-8e1f-c86fac276890.png)
![image](https://user-images.githubusercontent.com/34760505/146641555-d98e3c91-e533-4f3f-9633-cf4e0c9fafea.png)

**Comment by xhofe** [comment]
这个预览图是对应的存储策略提供的，如果有的可以支持。
对于其他的，封面和歌词播放器都可以支持显示，但是如何使用js提取是个问题，你有使用js提取内嵌封面或歌词推荐的库吗？

**Reply by ChirmyRam** [reply]
不太了解js，但是这里有一个h5播放器 [Aplayer](https://github.com/DIYgod/APlayer) ，不知能否参考它。

**Reply by xhofe** [reply]
我目前看过的播放器的封面图与歌词都是需要用户手动提供的，而不是在音乐文件中提取的，[Aplayer](https://github.com/DIYgod/APlayer)也是。所以得需要一个可以提取这些信息的库才可以，因为我也不知道如何使用js进行提取。

**Reply by ChirmyRam** [reply]
emmm那看来只有搁置这个功能了，搜了一圈有关内嵌歌词封面的信息很少。现在只能依靠其他支持WebDAV的本地播放器实现了。

另外能否加一个WebDAV支持读写权限的开关吗？我期望将WebDAV服务公开分享出来，所以担心读写操作权限，看了下目前OneDrive的WebDAV不支持读写，未来会有机会支持读写吗？如果不会支持我倒是不担心这一点，如果支持我希望有个权限开关，我本身倒也希望能有完整的读写权限。

**Reply by xhofe** [reply]
我找到一个 https://github.com/aadsm/jsmediatags 库似乎可以读取媒体属性信息，可以发我一份内嵌歌词与封面的文件测试吗

**Reply by xhofe** [reply]
对于WebDAV onedrive的写后续会添加，读写权限的开关我需要思考一下如何设计。

**Reply by ChirmyRam** [reply]
> 我找到一个 https://github.com/aadsm/jsmediatags 库似乎可以读取媒体属性信息，可以发我一份内嵌歌词与封面的文件测试吗

[Aimer - 春はゆく (春逝).mp3](https://al.chirmyram.com/d/rep/Music/ACG/Aimer%20-%20%E6%98%A5%E3%81%AF%E3%82%86%E3%81%8F%20(%E6%98%A5%E9%80%9D).mp3)
[AAA - With you (与你同在).flac](https://al.chirmyram.com/d/rep/Music/ACG/AAA%20-%20With%20you%20(%E4%B8%8E%E4%BD%A0%E5%90%8C%E5%9C%A8).flac)

提供了mp3、flac这两个格式，这两首均内置歌词和封面。

![image](https://user-images.githubusercontent.com/34760505/146645541-0b6ecdf7-26b7-47c8-894d-6a67535ad8f9.png)
![image](https://user-images.githubusercontent.com/34760505/146645557-404ff5aa-39b4-463f-901a-874e037f6b68.png)

**Reply by ChirmyRam** [reply]
> 对于WebDAV onedrive的写后续会添加，读写权限的开关我需要思考一下如何设计。

感谢。

**Reply by xhofe** [reply]
封面都可以获取到，但是flac的歌词似乎无法获取。
![image](https://user-images.githubusercontent.com/36558727/146646246-81db4b52-6ed9-4b64-a8f1-fcc3b5150223.png)

**Reply by ChirmyRam** [reply]
> 封面都可以获取到，但是flac的歌词似乎无法获取。 ![image](https://user-images.githubusercontent.com/36558727/146646246-81db4b52-6ed9-4b64-a8f1-fcc3b5150223.png)

我从它的issue中发现两个相关的库： [music-metadata-browser](https://github.com/Borewit/music-metadata-browser)  、 [music-metadata](https://github.com/Borewit/music-metadata) 看起来支持的格式挺多，不知道能否拿来用。

**Reply by xhofe** [reply]
放弃[jsmediatags](https://github.com/aadsm/jsmediatags)了，这个对[HTTP Access Control (CORS)](https://github.com/aadsm/jsmediatags#http-access-control-cors)要求太高了，几乎没有策略是支持`If-Modified-Since`的，我试试[music-metadata-browser](https://github.com/Borewit/music-metadata-browser)

**Reply by wc7086** [reply]
现在部分 flac 的封面获取不到，例如 #423  ，还有 WAV 文件的封面获取不到 例如 https://transfer.sh/OePSl5/%E6%89%8B%E5%AF%AB%E7%9A%84%E5%BE%9E%E5%89%8D.wav
